barman/postgres.py

```diff
@@ -1813,3 +1843,107 @@ def stop_exclusive_backup(self):
+
+    :cvar THREAD_NAME: The name identifying the keep-alive thread.
+    """
+
+    THREAD_NAME = "barman_keepalive_thread"
+
+    def __init__(self, postgres, interval, raise_exception=False):
+        """
+        Constructor.
+
+        :param barman.postgres.PostgreSQLConnection postgres: The
+            Postgres connection to keep alive.
+        :param int interval: An interval in seconds at which a
+            heartbeat query will be sent to keep the connection alive.
+            A value <= ``0`` won't start the keepalive.
+        :param bool raise_exception: A boolean indicating if an exception
+            should be raised in case the connection is lost. If ``True``, a
+            ``PostgresConnectionLost`` exception will be raised as soon as
+            it's noticed a connection failure. If ``False``, it will keep executing
+            normally until the context exits.
+        """
+        self.postgres = postgres
+        self.interval = interval
+        self.raise_exception = raise_exception
+        self._stop_thread = threading.Event()
+        self._thread = threading.Thread(
+            target=self._run_keep_alive,
+            name=self.THREAD_NAME,
+        )
+
+    def _prepare_signal_handler(self):
+        """
+        Set up a signal handler to raise an exception on the main thread when
+        the keep-alive thread wishes to interrupt it. This method listens for a
+        ``SIGUSR1`` signal and, when received, raises a ``PostgresConnectionLost``
+        exception.
+
+        .. note::
+            This code is, and only works if, executed while on the main thread. We
+            are not able to set a signal listener on a child thread, therefore this
+            method must be executed before the keep-alive thread starts.
+        """
+
+        def raise_exception(signum, frame):
+            raise PostgresConnectionLost("Connection to Postgres server was lost.")
+
+        signal.signal(signal.SIGUSR1, raise_exception)
+
+    def _raise_exception_on_main(self):
+        """
+        Trigger an exception on the main thread at whatever frame is being executed
+        at the moment. This is done by sending a ``SIGUSR1`` signal to the process,
+        which will be caught by the signal handler set previously in this class.
+
+        .. note::
+            This is an alternative way of interrupting the main thread's work, since
+            there is no direct way of killing or raising exceptions on the main thread
+            from a child thread in Python. A handler for this signal has been set
+            beforehand by the ``_prepare_signal_handler`` method in this class.
+        """
+        os.kill(os.getpid(), signal.SIGUSR1)
+
+    def _run_keep_alive(self):
+        """Runs the keepalive until a stop-thread event is set"""
+        while not self._stop_thread.is_set():
+            if not self.postgres.has_connection:
+                # Wait for the connection to be opened on the main thread
```

Comment:
I wonder if we should have a very short sleep time here before going to the next loop iteration?
I mean, just so we avoid running "continue" statement several times waiting for the connection to be ready.

Reply:
1s is enough